code/parachain/runtime/dali/src/migrations.rs

```diff
@@ -19,6 +21,602 @@ impl OnRuntimeUpgrade for SchedulerMigrationV1toV4 {
+	}
+
+	impl AssetCreationInput {
+		fn new_asset(
+			asset_id: CurrencyId,
+			location: Option<XcmAssetLocation>,
+			name: Option<BiBoundedAssetName>,
+			symbol: Option<BiBoundedAssetSymbol>,
+			decimals: u8,
+			existential_deposit: Balance,
+			ratio: Option<Rational64>,
+		) -> Self {
+			Self {
+				asset_id,
+				location,
+				asset_info: AssetInfo { name, symbol, decimals, existential_deposit, ratio },
+			}
+		}
+	}
+
+	fn asset_info_update(asset_info: AssetInfo<Balance>) -> AssetInfoUpdate<Balance> {
+		AssetInfoUpdate {
+			name: Some(asset_info.name),
+			symbol: Some(asset_info.symbol),
+			decimals: Some(asset_info.decimals),
+			existential_deposit: Some(asset_info.existential_deposit),
+			ratio: Some(asset_info.ratio),
+		}
+	}
+
+	fn add_assets_to_storage(assets: Vec<AssetCreationInput>) -> Weight {
+		let (mut count_created, mut count_updated) = (0, 0);
+		for asset_input in assets {
+			let AssetCreationInput { asset_id, location, asset_info } = asset_input;
+			// check if there is data stored for foreign asset
+			if let Some(foreign_location) = location.clone() {
+				// check that new asset_id is the same as old one for the same location
+				let mut location_stored = false;
+				if let Some(prev_asset_id) =
+					<AssetsRegistry as RemoteAssetRegistryInspect>::location_to_asset(
+						foreign_location.clone(),
+					) {
+					if prev_asset_id != asset_id {
+						panic!("previous and new asset_id for location do not match");
+					}
+					location_stored = true;
+				}
+				// check that new location is the same as old one for the same asset_id
+				let mut asset_stored = false;
+				if let Some(prev_location) =
+					<AssetsRegistry as RemoteAssetRegistryInspect>::asset_to_remote(asset_id)
+				{
+					if prev_location != foreign_location {
+						panic!("previous and new location for asset_id do not match");
+					}
+					asset_stored = true;
+				}
+				// check that either both maps or none map asset_id and location
+				if location_stored != asset_stored {
+					panic!("ForeignToLocal and LocalToForeign maps contradict each other");
+				}
+				if location_stored {
+					<AssetsRegistry as RemoteAssetRegistryMutate>::update_asset(
+						asset_id,
+						asset_info_update(asset_info.clone()),
+					)
+					.expect("asset wasnt updated");
+					count_updated += 1;
+				} else {
+					<AssetsRegistry as RemoteAssetRegistryMutate>::register_asset(
+						asset_id,
+						location.clone(),
+						asset_info.clone(),
+					)
+					.expect("asset wasnt registered");
+					count_created += 1;
+				}
+				continue
+			}
+			// check that for local asset there is no location previously stored
+			if let Some(_prev_location) =
+				<AssetsRegistry as RemoteAssetRegistryInspect>::asset_to_remote(asset_id)
+			{
+				panic!("location is not None for local asset_id");
+			}
+			// check if there is local asset with asset_id
+			if <AssetsRegistry as InspectRegistryMetadata>::asset_name(&asset_id).is_some() {
```

Comment:
`asset_name` is intentionally optional in storage, meaning that an asset may exists with the ID but may not have a name.
I would instead rely on the `existential_deposit` field being `Some` instead - as all assets registered in the new system must have an existential deposit.